### PR TITLE
feat(PipelineProvider): add methods for removing pipeline steps

### PR DESCRIPTION
**Pull request by z-brooks**
https://github.com/aurelia/router/issues/332

### CODE REVIEW

**Comment by EisenbergEffect** [comment]
@bryanrsmith Can you review this PR?

File: src/pipeline-provider.js [hunk 1]
```diff
@@ -43,34 +58,55 @@ export class PipelineProvider {
...
-      static slotAlias = alias;
-      static steps = [];
-
-      constructor(container) {
-        this.container = container;
-      }
+  /**
+   * Removes a step from a slot in the pipeline
+   */
+  removeStep( name:string, step: PipelineStep ) {
```
Comment (bryanrsmith) [inline src/pipeline-provider.js]: If you could match the existing style by using a single space after colons, and no space inside parens that would be great.

File: src/pipeline-provider.js [hunk 2]
```diff
@@ -43,34 +58,55 @@ export class PipelineProvider {
...
+    }
+  }
 
-      getSteps() {
-        return PipelineSlot.steps.map(x => this.container.get(x));
-      }
+  /**
+   * Clears all steps from a slot in the pipeline
+   */
+  clearSteps( name:string = '' ) {
```
Comment (bryanrsmith) [inline src/pipeline-provider.js]: Should this be private? It seems like a dangerous thing to allow.

**Comment by bryanrsmith** [comment]
Left a couple comments, but otherwise looks good to me.

**Comment by EisenbergEffect** [comment]
@z-brooks Could you address the comments that @bryanrsmith left? If so, I think we could get this in.

**Comment by z-brooks** [comment]
Changes made.

**Comment by EisenbergEffect** [comment]
Thanks! @jedd-ahyoung or @bryanrsmith Can you do a final review?

**Comment by jedd-ahyoung** [comment]
@EisenbergEffect I checked this out.

@z-brooks This actually looks quite good, but I think that the class declaration here (https://github.com/aurelia/router/pull/334/files#diff-0b0b3e28fdaaf40a0b48ac8397c14380R14) needs to have the static properties defined in order for the factory (https://github.com/aurelia/router/blob/master/src%2Fpipeline-provider.js) to work. Placing the properties inside the constructor (i.e. the instance) means that they won't be available until the slots are instantiated. Have you verified that the tests are passing and that this is working with the skeleton-navigation app?

Additionally, we need to ensure that we have a `static inject` for the container instance.

Other than those issues, this looks good. (One of the problems here is that we have a lack of tests to catch this stuff.)

**Comment by z-brooks** [comment]
@jedd-ahyoung All tests pass and it works fine in a skeleton. I'm also using this in my project (noted in the linked issue) without issue.

Not sure what you mean by a static inject for the container instance. `Container` is injected into the `PipelineProvider` [here](https://github.com/z-brooks/router/blob/pipeline-provider/src/pipeline-provider.js#L32) and I just pass it when constructing a new slot [here](https://github.com/z-brooks/router/blob/pipeline-provider/src/pipeline-provider.js#L110).

**Comment by EisenbergEffect** [comment]
This is odd because when the provider loops through the steps it then tries to resolve an *existing instance* of PipelineStep from the container. However, it does work because when trying to resolve an object instance without a pre-registration the object instance will resolve to itself. So, it does work... We might want to put some comments to this affect here.

**Pull request merged**